File: webflux/src/main/java/org/sample/webflux/component/RequestValidationExceptionHandler.java
package org.sample.webflux.component;

import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.support.WebExchangeBindException;
import reactor.core.publisher.Mono;

import jakarta.validation.ConstraintViolationException;

@Slf4j
@ControllerAdvice
public class RequestValidationExceptionHandler {

    @ExceptionHandler(
            value = { WebExchangeBindException.class }
    )
    protected Mono<ResponseEntity<String>> handleWebExchangeBindException(
            WebExchangeBindException ex
    ) {
        log.debug("request validation error:", ex);
        return Mono.just(ResponseEntity.badRequest().body(ex.getMessage()));
    }

    @ExceptionHandler(
            value = { ConstraintViolationException.class }
    )
    protected Mono<ResponseEntity<String>> handleConstraintViolation(
            ConstraintViolationException ex
    ) {
        log.debug("request validation error:", ex);
        return Mono.just(ResponseEntity.badRequest().body(ex.getMessage()));
    }
}

File: webflux-kt/src/main/kotlin/org/sample/webflux/component/RequestValidationExceptionHandler.kt
package org.sample.webflux.component

import io.klogging.NoCoLogging
import jakarta.validation.ConstraintViolationException
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.support.WebExchangeBindException

@ControllerAdvice
class RequestValidationExceptionHandler : NoCoLogging {
    @ExceptionHandler(value = [WebExchangeBindException::class])
    fun handleWebExchangeBindException(ex: WebExchangeBindException): ResponseEntity<String> {
        logger.debug("web exchange bind error:", ex)
        return ResponseEntity.badRequest().body(ex.message)
    }

    @ExceptionHandler(value = [ConstraintViolationException::class])
    fun handleConstraintViolation(ex: ConstraintViolationException): ResponseEntity<String> {
        logger.debug("constraint violation error:", ex)
        return ResponseEntity.badRequest().body(ex.message)

    }
}
